[DO NOT MERGE] Improved generic translation

mrtcode opened the pull request.

mrtcode committed:
Improved generic translation

mvolz commented on Generic.js:
```diff
@@ -0,0 +1,499 @@
+}
+
+/** BEGIN TEST CASES **/
+var testCases = [];
```
It's be helpful to have test cases here.

zuphilip commented:
What is the status/plan for this PR which relies on the new functions for JSON-LD in `ZU` https://github.com/zotero/zotero/pull/1656 ? CC @mrtcode @dstillman

AbeJellinek committed:
normalize -> asciify

AbeJellinek committed:
Generic: Update DOI translator ID

AbeJellinek committed:
Generic: Use non-"webpage" item type when found

AbeJellinek committed:
Generic: Only use head DOIs when found?

Not sure about this, but headDois wasn't being used at all before.

AbeJellinek committed:
Generic: Add all tests from EM, plus one for item type change

AbeJellinek committed:
Update tests from EM

AbeJellinek committed:
Look for elements with 'heading' ARIA role

AbeJellinek committed:
Progress toward detection

EM tends to detect 'webpage' for non-webpage items, so this doesn't work
perfectly yet.

AbeJellinek committed:
More progress toward detection

- Don't detect in generic translators unless called from Generic
- Fix detection code
- Fix multiples code
- Add logging
- Don't use Zotero.done() in EM
- Detect more accurately in LM

mvolz commented:
> OK, this is working well enough to test with [zotero/zotero#2894](https://github.com/zotero/zotero/pull/2894) and [zotero/utilities#20](https://github.com/zotero/utilities/pull/20) checked out.

Great! Adding json-ld seems it would help a lot, see this example where the json-ld was much better than using the byline: https://phabricator.wikimedia.org/T432076
